nats-base-client/kv.ts

```diff
@@ -0,0 +1,274 @@
+  delete(k: string): Promise<void>;
+  purge(opts?: PurgeOpts): Promise<PurgeResponse>;
+  destroy(): Promise<boolean>;
+}
+
+export class Bucket implements KV {
+  jsm: JetStreamManager;
+  js: JetStreamClient;
+  bucket: string;
+  maxPerSubject!: number;
+
+  constructor(bucket: string, jsm: JetStreamManager, js: JetStreamClient) {
+    this.jsm = jsm;
+    this.js = js;
+    this.bucket = bucket;
+  }
+
+  async init(opts: Partial<BucketOpts> = {}): Promise<void> {
+    const sc = {} as StreamConfig;
+    sc.name = opts.streamName ?? this.bucketName();
+    sc.subjects = [this.subjectForBucket()];
+    sc.retention = RetentionPolicy.Limits;
+    this.maxPerSubject = sc.max_msgs_per_subject = opts.history ?? 5;
+    sc.max_msgs = -1;
+    sc.storage = StorageType.File;
+    sc.discard = DiscardPolicy.Old;
+    sc.num_replicas = opts.replicas ?? 1;
+
+    try {
+      await this.jsm.streams.info(sc.name);
+    } catch (err) {
+      if (err.message === "stream not found") {
+        await this.jsm.streams.add(sc);
+      }
+    }
+  }
+
+  bucketName(): string {
+    return `${kvPrefix}${this.bucket}`;
+  }
+
+  subjectForBucket(): string {
+    return `${kvSubjectPrefix}.${this.bucket}.*`;
+  }
+
+  subjectForKey(k: string): string {
+    return `${kvSubjectPrefix}.${this.bucket}.${k}`;
+  }
+
+  static async create(
+    nc: NatsConnection,
+    name: string,
+    opts: Partial<BucketOpts> = {},
+  ): Promise<KV> {
+    const to = opts.timeout || 2000;
+    const jsm = await nc.jetstreamManager({ timeout: to });
+    const bucket = new Bucket(name, jsm, nc.jetstream({ timeout: to }));
+    await bucket.init(opts);
+    return bucket;
+  }
+
+  close(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  async delete(k: string): Promise<void> {
+    const ji = this.js as JetStreamClientImpl;
+    const cluster = ji.nc.info?.cluster ?? "";
+    const h = headers();
+    h.set(kvOriginClusterHdr, cluster);
+    h.set(kvOperationHdr, "DEL");
+    await this.js.publish(this.subjectForKey(k));
+  }
+
+  toResult(k: string, sm: StoredMsg): Result {
+    return {
+      bucket: this.bucket,
+      key: k,
+      data: sm.data,
+      created: sm.time,
+      sequence: sm.seq,
+      origin_cluster: sm.header.get(kvOriginClusterHdr),
+      operation: sm.header.get(kvOperationHdr) === "DEL" ? "DEL" : "PUT",
+    };
+  }
+
+  jtoResult(k: string, jm: JsMsg): Result {
+    return {
+      bucket: this.bucket,
+      key: k,
+      data: jm.data,
+      created: new Date(millis(jm.info.timestampNanos)),
+      sequence: jm.seq,
+      origin_cluster: jm.headers?.get(kvOriginClusterHdr),
+      operation: jm.headers?.get(kvOperationHdr) === "DEL" ? "DEL" : "PUT",
+    };
+  }
+
+  async get(k: string): Promise<Result> {
+    const sm = await this.jsm.streams.getMessage(this.bucketName(), {
+      last_by_subj: this.subjectForKey(k),
+    });
+    return this.toResult(k, sm);
+  }
+
+  async history(k: string): Promise<Result[]> {
```

Comment:
still not really sure how I want to expose history to clients so wasnt in the ADR iirc, but this seems fine anyway

Reply:
I want it for ngs